File: lld/6.py
class Database:
    def __init__(self, host, port, ssl, pool_size):
        self.__host = host
        self.__port = port
        self.__ssl = ssl
        self.__pool_size = pool_size

    def connect(self):
        pass


class DatabaseBuilder:
    def __init__(self):
        self.__host = None
        self.__port = 5432  # default
        self.__ssl = False
        self.__pool_size = 10

    def with_host(self, host: str) -> "DatabaseBuilder":
        self.__host = host
        return self

    def with_port(self, port: int) -> "DatabaseBuilder":
        self.__port = port
        return self

    def with_ssl(self, ssl: bool) -> "DatabaseBuilder":
        self.__ssl = ssl
        return self

    def with_pool_size(self, size: int) -> "DatabaseBuilder":
        self.__pool_size = size
        return self

    def build(self) -> Database:
        if not self.__host:
            raise ValueError("host is required")
        return Database(self.__host, self.__port, self.__ssl, self.__pool_size)


# usage
# INFO: main is the that how with funtion return the serlf "DatabaseBuilder"
db = DatabaseBuilder().with_host("localhost").with_ssl(True).with_pool_size(100).build()
# ```
#
# ---
#
# ## The key insight — `return self`
#
# Every `With` method returns the builder itself. That's what enables chaining:
# ```
# NewDatabase()          → returns builder
# .WithHost("localhost") → sets host, returns same builder
# .WithPort(5432)        → sets port, returns same builder
# .Build()               → validates, constructs, returns final object

File: lld/6.go
package main

import (
	"errors"
	"fmt"
	"time"
)

type DatabaseBuilder struct {
	host       string
	port       int
	ssl        bool
	poolSize   int
	timeout    time.Duration
	maxRetries int
}

func NewDatabase() *DatabaseBuilder {
	return &DatabaseBuilder{
		port:       5432, // sensible defaults
		poolSize:   10,
		timeout:    30 * time.Second,
		maxRetries: 3,
	}
}

// each method sets one field and returns the builder
func (b *DatabaseBuilder) WithHost(host string) *DatabaseBuilder {
	b.host = host
	return b // INFO: returns self for chaining
}

func (b *DatabaseBuilder) WithPort(port int) *DatabaseBuilder {
	b.port = port
	return b
}

func (b *DatabaseBuilder) WithSSL(ssl bool) *DatabaseBuilder {
	b.ssl = ssl
	return b
}

func (b *DatabaseBuilder) WithPoolSize(size int) *DatabaseBuilder {
	b.poolSize = size
	return b
}

// Build validates and constructs the final object
func (b *DatabaseBuilder) Build() (*pgxpool.Pool, error) {
	if b.host == "" {
		return nil, errors.New("host is required")
	}
	connStr := fmt.Sprintf(
		"postgres://%s:%d?sslmode=%v&pool_max_conns=%d",
		b.host, b.port, b.ssl, b.poolSize,
	)
	return pgxpool.New(context.Background(), connStr)
}
